File: problem-solving/lc-1-easy-two-sum/2-optimized.py
"""
Problem: https://leetcode.com/problems/two-sum/

Complexity
----------
Time: O(N)
Space: O(N)
"""

class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        seen = {}
        for i in range(len(nums)):
            if target - nums[i] in seen:
                return [i, seen[target - nums[i]]]
            seen[nums[i]] = i

File: problem-solving/lc-1-easy-two-sum/2-optimized.go
/*
Problem: https://leetcode.com/problems/two-sum/

Complexity
----------
Time: O(N)
Space: O(N)
*/

func twoSum(nums []int, target int) []int {
    seen := make(map[int]int)
    for i, n := range nums {
        if idx, ok := seen[target - n]; ok {
            return []int{i, idx}
        }
        seen[n] = i
    }
    return []int{}
}
